app/lua53/lundump.c

```diff
@@ -1,637 +1,876 @@
+** Unlike the standard dump usage, this writer returns a void * value which
+** is the flash location written to.
 */
-static void LoadProtos (LoadState *S, Proto *f) {
-  int i, n = LoadInt(S);
-  Proto **p = luaM_newvector(S->L, n, Proto *);
-  f->p = p;
-  f->sizep = n;
-  memset (p, 0, n * sizeof(*p));
-  for (i = 0; i < n; i++)
-    p[i] = LoadFunction(S, luaF_newproto(S->L), f->source);
-  if (S->mode != MODE_RAM) {
-    f->p = StoreAV(S, cast(void **, p), n);
-    luaM_freearray(S->L, p, n);
-  }
-}
-static void LoadUpvalues (LoadState *S, Proto *f) {
-  int i, nostripnames = LoadByte(S);
-  f->sizeupvalues = LoadInt(S);
-  if (f->sizeupvalues) {
-    f->upvalues = NewVector(S, f->sizeupvalues, Upvaldesc);
-    for (i = 0; i < f->sizeupvalues ; i++) {
-      TString *name = nostripnames ? LoadString(S) : NULL;
-      Upvaldesc uv = {name, LoadByte(S), LoadByte(S)};
-      StoreR(S, f->upvalues, i, uv, FMT_UPVALUE);
+
+#define FLASH_WRITE(S,b,l) S->flashWriter(S->flashWriterData, b, l)
+#define FLASH_SYNC(S) FLASH_WRITE(S, NULL, ~(size_t)0)
+#define FLASH_CURR(S) FLASH_WRITE(S, NULL, 0)
+
+/*
+** The following compression maps are used in Store() ONLY in the case of
+** 64-bit hosts to repack structs to adjust for 64-bit->32-bit points in
+** absolute mode LFS creation.  These must match the definitions in lobject.h
+*/
+typedef Proto *pProto;
+#ifdef LUA_USE_HOST64
+# define FMT_DECL   , const char *pack
+# define FMT_PACK   , pack
+# define FMT_ARG(t) , FMT_ ## t
+# define LFS_ABS_MODE(S)  (S->mode == MODE_LFSA)
+# define FMT_LocVar        "Aww"
+# define FMT_NULL          NULL
+# define FMT_Instruction   "w"
+# define FMT_Proto         "AwwwwwwwwwwAAAAAAAAA"
+# define FMT_pProto        "A"
+# define FMT_ROTable       "AWAA"
+# define FMT_ROTable_entry "AWA"
+# define FMT_TString       "AwwA"
+# define FMT_UTString      ((sizeof(TString) == sizeof(UTString) || \
+                             S->mode == MODE_LFSA) ? FMT_TString : \
+                                                     FMT_TString "SSSSS")
+# define FMT_pTString      "A"
+# define FMT_TValue        "WA"
+# define FMT_Upvaldesc     "AW"
+#else
+# define FMT_DECL
+# define FMT_PACK
+# define FMT_ARG(t)
+# define LFS_ABS_MODE(S)  (0)
+#endif
+
+static void *Store (LoadState *S,
+                    void *vec, int ndx,     /* dest is vec[ndx] in len chunks */
+                    const void *src, size_t len           /* src in len chunk */
+                    FMT_DECL) {   /* One of above pack formats on HOST builds */
+  lu_byte *dest;                              /* returns addr of dest element */
+  if (inRAM(S)) {
+    dest = cast(lu_byte *, vec) + (ndx*len);
+    if (dest != src)
+      memcpy(dest, src, len);
+#ifdef LUA_USE_HOST64
+  } else if (S->mode == MODE_LFSA && pack) {
+   /*
+    * 64-bit hosts have two LFS modes. The default LFS mode is for writing to a
+    * when writing to native LFS where the address pointers are 64 bit and any
```

Comment:
Is there a line missing here? Or is it just a mangled comment?

Reply:
mangled will rewrite